Lexer

nitnelave opened the pull request.

nitnelave committed:
make: switch to debug mode by default

Berdes commented on src/error/error.h:
```diff
@@ -0,0 +1,161 @@
+class MaybeError {
+ public:
+  // Constructor for no error.
+  // Not explicit on purpose.
```
If a constructor doesn't have any argument, it cannot be explicit anyway.

Berdes commented on src/error/error.h:
```diff
@@ -0,0 +1,161 @@
+// Either an error or a positive result.
+// This class is immutable.
+template <typename Err>
+class MaybeError {
```
Why are you storing the contained error directly in the class instead of just using a pointer? For me that sounds overly complex and very error-prone.

The obvious solution would be to have a unique_ptr<Err> that you either construct with a nullptr or by moving/copying the contained error. This would also have the advantage of not wasting memory for a potential error when there is none.

Berdes commented on src/error/error.h:
```diff
@@ -0,0 +1,161 @@
+  // Return the error if it is one, otherwise return "Ok".
+  std::string toString() const {
+    if (ok())
+      return error_or_die().toString();
```
You reversed the two cases, you should return "Ok" in this case.

Berdes commented on src/error/error.h:
```diff
@@ -0,0 +1,161 @@
+// Either an error or a T.
+// This class is immutable (but the T can be mutable).
+template <typename T, typename Err = Error>
+class ErrorOr {
```
Same as above: why not two unique_ptr to store the error and value?

Berdes commented on src/error/error.h:
```diff
@@ -0,0 +1,161 @@
+class ErrorOr {
+ public:
+  // Not explicit on purpose.
+  ErrorOr(Err&& error) : is_ok_(false) {
```
You provide a constructor to copy error in MaybeError, you should probably also add it here.

Berdes reviewed (changes requested):
Sorry, I thought the review was per-commit and not for the whole batch.

Berdes commented on src/lexer/lexer.cc:
```diff
@@ -20,11 +20,31 @@ std::unique_ptr<std::istream> get_stream_from_source(const std::string& source,
+    }
+    ++location_.column;
+    if (!stream_->get(next_char_))
+        return makeError("Could not read from stream while parsing at position ");
```
You don't need the "at position" since it's already added in the constructor of LexerError.

Berdes commented on src/lexer/lexer.cc:
```diff
@@ -20,11 +20,31 @@ std::unique_ptr<std::istream> get_stream_from_source(const std::string& source,
+
+ErrorOr<Token, LexError> Lexer::get_next_token() {
+  if (location_.line == 0)
+    return LexError("Invalid location", location_);
```
You could use makeError since you are using the current location.

Berdes commented on src/lexer/lexer.h:
```diff
@@ -5,30 +5,43 @@
+  // Puts the value of the next char into next_char_, and updates the location.
+  MaybeError<LexError> get_next_char();
+
+  char next_char_ = 0;
```
Do you plan for the language to be purely ASCII? If you think it will be UTF-8 (like some reasonable languages), this will not work.

Berdes commented on src/parser/parser.cc:
```diff
@@ -7,12 +7,15 @@ using lexer::TokenType;
+MaybeError<ParseError> Parser::parse() {
+  auto res = lexer_->get_next_token();
+  if (!res.ok())
+    return {res.error_or_die()};
```
This kind of construct (get result and propagate in case of error) will certainly be common, what about creating a macro for it? This would avoid some potential stupid errors (like forgetting the ! in the condition).

Berdes commented on src/parser/parser.h:
```diff
@@ -1,16 +1,25 @@
+  static std::string location_line(const std::string& message,
+                                   const lexer::Token& t) {
+    std::stringstream ss;
+    ss << message << "\n At " << t.begin.toString();
```
Maybe it's better to indicates "Between {begin} and {end}". In which case, you probably want to avoid the repetition of the file name.

nitnelave committed:
clang-tidy: disable questionable checks

nitnelave committed:
travis,cmake: clean up

nitnelave committed:
error: add printing facilities

nitnelave committed:
tools: clean up clang-format

nitnelave committed:
lexer: enrich token class

nitnelave committed:
lexer: handle basic tokens

nitnelave committed:
test: add utilities

nitnelave committed:
lexer: add tests

nitnelave committed:
lexer: make unget_char update next_char_

nitnelave committed:
lexer: document

nitnelave committed:
lexer: add support for identifiers and keywords

nitnelave committed:
tokens: fix X-macro conflict, add keyword bounds

X-macro keyword was conflicting with keyword "DO".

nitnelave committed:
tests: rename compare_operator function

It is not restricted to operators, but to tokens with identical symbolic
representation.

nitnelave committed:
tests: add identifier lexing tests

nitnelave committed:
tests: add location tests for identifiers and comments

nitnelave committed:
tools: add missing double quotes in clang-format.sh
